### DISCUSSION TITLE
DAI 2023-2024 - Java IOs benchmarking - Alexandre Iorio

### DISCUSSION THREAD

**Discussion by AlexandreIorio** [heig-vd-dai-course/heig-vd-dai-course#75] [Show and tell]
[https://github.com/AlexandreIorio/DAI-java-ios](https://github.com/AlexandreIorio/DAI-java-ios)

![Chart](https://raw.githubusercontent.com/AlexandreIorio/DAI-java-ios/98192f74f2cbc75cba38f888abd99d07b338344a/chart.svg)
![DataTable](https://raw.githubusercontent.com/AlexandreIorio/DAI-java-ios/98192f74f2cbc75cba38f888abd99d07b338344a/data_table.svg)

#### Which type of stream is the most efficient for each use case?

Binary streams are the most efficient for all use cases, regardless of whether or not buffering is enabled.

#### Why is it more efficient than the other types of streams?

Binary streams are more efficient because they operate directly on bytes, without having to perform any encoding or decoding. This is especially important when dealing with large amounts of data.

Text streams, on the other hand, must encode and decode the data each time it is read or written. This can add a significant overhead, especially when dealing with large files.

#### What is the difference between binary data and text data?

Binary data is data that is stored in its native format, without any encoding or decoding. This is the most efficient way to store data, but it is not human-readable.

Text data, on the other hand, is data that has been encoded to make it human-readable. This is less efficient than storing binary data, but it is much easier to read and write.

#### What is a character encoding?

A character encoding is a way of representing characters as a sequence of bytes. There are many different character encodings, each with its own advantages and disadvantages.

The most common character encoding is ASCII, which uses 7 bits to represent each character. This allows ASCII to represent 128 different characters, including all of the letters, numbers, and punctuation symbols that are commonly used in English.

However, ASCII cannot represent all of the characters that are used in other languages. For example, it cannot represent the characters that are used in Chinese, Japanese, or Korean.

To represent these characters, other character encodings are used, such as UTF-8. UTF-8 uses a variable number of bytes to represent each character, which allows it to represent a much wider range of characters than ASCII.

#### Why is this methodology important?

It is important to understand the difference between binary and text streams, and to choose the right type of stream for each use case. This can help to improve the performance and efficiency of your application.

For example, if you are storing data that does not need to be human-readable, such as images or audio files, you should use a binary stream. This will allow you to store the data in the most efficient way possible.

However, if you are storing data that needs to be human-readable, such as text files or configuration files, you should use a text stream. This will make the data easier to read and write.

In addition to choosing the right type of stream, it is also important to consider whether or not to enable buffering. Buffering can improve the performance of your application by reducing the number of times that the underlying I/O system needs to be accessed.

However, buffering can also add a small overhead, so it is important to weigh the benefits and drawbacks before enabling it.

Overall, understanding the difference between binary and text streams, and choosing the right type of stream for each use case, is an important part of writing efficient Java code.

*information search assistance: Bard text-based IA Google*

**Comment by ludelafo** [comment]
Great benchmark methodology!

> Binary streams are the most efficient for all use cases, regardless of whether or not buffering is enabled.

While totally true when discussing the binary/text performance, it's still important to use the right classes when working with text files to avoid misinterpretation of the character encoding. Even if binary streams would be more performant, it is not guaranteed that the interpretation will be made correctly if you have to analyze the words from the string for example. Buffered streams are always recommended for the best performance, as it is very rare to work **only** with files made of a few bytes. Using buffered streams ensure the best performance in most use-cases.

In your codebase, for text files, do not forget to specify the charset of the file (while reading and writing!). It is more explicit and you can avoid many issues in the future.

One of the constructors for the `FileReader` et `FileWriter` classes allows to specify the charset to ensure the characters are interpreted correctly by your application: 

- https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/FileReader.html
- https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/FileWriter.html

**EDIT**: Thank you for citing the search assistance! Great usage of the tool. Please keep in mind that we cannot allow these tools during the evaluations and the exam and always keep a critical mind with the answer it provides.
